src/library/scala/collection/immutable/ChampHashMap.scala

```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
   }
 
   override def concat[V1 >: V](that: scala.IterableOnce[(K, V1)]): HashMap[K, V1] = {
```

Comment:
I'm assuming that when we [optimize merge](https://github.com/scala/bug/issues/11077) we'll be able to share the implementation with `concat`. Do you think it's worth doing `merge` in this PR so we can shake out a good structure?

Reply:
I actually don't think we'll be able to share implementation with `concat` -- the result of the `merge` function is not constrained to `leftKey | rightKey`, it can be any `K`, so can fall anywhere in the resulting map's trie structure.